specs/emip-005.md

```diff
@@ -0,0 +1,472 @@
+
+Note: `payload` can be `nil`. This means that the payload is known by both the signer and the verifier and therefore doesn't need to be encoded.
+
+For your convenience, the structure is provided here:
+
+```
+empty_or_serialized_map = bstr .cbor header_map / bstr .size 0
+
+header_map = {
+    Generic_Headers,   # reserved headers (see section 3.1)
+    * label => values  # any number of string labels for application-specific purpose.
+}
+
+Headers = (
+    protected : empty_or_serialized_map,
+    unprotected : header_map
+)
+
+# signature layer
+COSE_Signature =  [
+    Headers,
+    signature : bstr
+]
+
+# if signing with just ONE key
+COSE_Sign1 = [
+    Headers,
+    payload : bstr / nil,
+    signature : bstr
+]
+
+# if signing with >1 key
+COSE_Sign = [
+    Headers,
+    payload : bstr / nil,
+    signatures : [+ COSE_Signature]
+]
+
+SIGN_TAG = 18 | 98   # unique prefix so we can infer which structure is used
+signed_message = 98 ( COSE_SIGN) / 18 ( COSE_Sign1)
+```
+
+### Signing and Verification target format
+
+Instead of signing the full structure, we instead sign the following type which is derived from the structure
+
+```
+Sig_structure = SIGN_TAG ([
+  context : "Signature" / "Signature1" / "CounterSignature",    # explained below
+  body_protected : empty_or_serialized_map,                     # protected from layer 1
+  ? sign_protected : empty_or_serialized_map,                   # not present in Sign1 case
+  external_aad : bstr,                                          # explanation below
+  payload : bstr
+])
+```
+
+⚠️ Note: We add a `SIGN_TAG` not present inside `Sig_structure` in the original COSE specification. This is because we want to avoid a user accidentally signing a byte sequence that may be confused as a different structure with different meaning (notably, we want to avoid the user accidentally signing a transaction or a block). The way we do this is that Cardano enforces that every signable object MUST start with a unique prefix. You can find a current list in the Rust codebase [here](https://github.com/input-output-hk/rust-cardano/blob/2e924a4457203fa6b19f409cb897d90cd6b699f6/cardano/src/tags.rs) and the Haskell codebase [here](https://github.com/input-output-hk/cardano-sl/blob/ccca8b346413f12e9fad8cab4c855e69669a2151/crypto/Pos/Crypto/Signing/Tag.hs#L15).
+
+We therefore introduce a set of new tags based off `Table 1` in [RFC 8152](https://tools.ietf.org/html/rfc8152).
+
+| CBOR Tag | Type          |
+|----------|---------------|
+| 98       | COSE_Sign     |
+| 18       | COSE_Sign1    |
+| 96       | COSE_Encrypt  |
+| 16       | COSE_Encrypt0 |
+| 97       | COSE_Mac      |
+| 17       | COSE_Mac0     |
+
+Note that we introduce tags that are not used for this specific specification. This is just to reserve them in case they are used in the future.
```

Comment:
... specific specification ... tautology. Probably, it may be "distinct specification"